File: archive/flutter-app/test/core/utility/gzip_converter_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:infrakit_studio/core/utility/gzip_converter.dart';

void main() {
  const converter = GzipConverter();

  group('compress / decompress round trip', () {
    test('text survives a gzip round trip unchanged', () {
      const original = 'InfraKit Studio — gzip round trip with ünïcode ✓\nline two\n';

      final compressed = converter.compressText(original);
      final restored = converter.decompressBytes(compressed.data);

      expect(restored.textOrNull, original);
      expect(restored.originalSize, utf8.encode(original).length);
    });

    test('arbitrary binary bytes survive a round trip unchanged', () {
      final bytes = Uint8List.fromList([for (var i = 0; i < 512; i++) (i * 37) % 256]);

      final compressed = converter.compressBytes(bytes);
      final restored = converter.decompressBytes(compressed.data);

      expect(restored.data, bytes);
    });

    test('gzip output starts with the gzip magic number', () {
      final compressed = converter.compressText('hello');
      expect(GzipConverter.looksLikeGzip(compressed.data), isTrue);
      expect(compressed.data[0], 0x1f);
      expect(compressed.data[1], 0x8b);
    });
  });

  group('size reporting', () {
    test('a highly repetitive string compresses to a fraction of its size', () {
      final repetitive = 'the same line over and over\n' * 400;

      final result = converter.compressText(repetitive);

      expect(result.originalSize, utf8.encode(repetitive).length);
      expect(result.compressedSize, lessThan(result.originalSize ~/ 10));
      expect(result.compressionRatio, lessThan(0.1));
      expect(result.spaceSavingPercent, greaterThan(90));
    });

    test('sizes are reported from the payload side regardless of direction', () {
      final compressed = converter.compressText('a' * 1000);
      final restored = converter.decompressBytes(compressed.data);

      expect(restored.originalSize, 1000);
      expect(restored.compressedSize, compressed.compressedSize);
      expect(restored.compressionRatio, closeTo(compressed.compressionRatio, 1e-9));
    });
  });

  group('base64 transport', () {
    test('text -> gzip -> base64 -> gzip -> text round trips', () {
      const original = 'kubectl get pods -A | grep CrashLoopBackOff';

      final compressed = converter.compressText(original);
      final restored = converter.decompressBase64(compressed.base64);

      expect(restored.textOrNull, original);
    });

    test('base64 pasted with line breaks and no padding still decodes', () {
      final compressed = converter.compressText('wrapped payload ' * 20);

      final wrapped = _wrap(compressed.base64, 40);
      final unpadded = wrapped.replaceAll('=', '');

      expect(converter.decompressBase64(wrapped).textOrNull, 'wrapped payload ' * 20);
      expect(converter.decompressBase64(unpadded).textOrNull, 'wrapped payload ' * 20);
    });

    test('url-safe base64 alphabet is accepted', () {
      final compressed = converter.compressText('url safe alphabet test payload');
      final urlSafe = compressed.base64.replaceAll('+', '-').replaceAll('/', '_');

      expect(converter.decompressBase64(urlSafe).textOrNull, 'url safe alphabet test payload');
    });
  });

  group('error handling', () {
    test('decompressing non-gzip bytes throws a clean FormatException', () {
      expect(
        () => converter.decompressBytes(Uint8List.fromList(utf8.encode('this is just text'))),
        throwsA(
          isA<FormatException>().having((e) => e.message, 'message', contains('Not a GZip stream')),
        ),
      );
    });

    test('decompressing a truncated gzip stream throws a clean FormatException', () {
      final compressed = converter.compressText('truncate me ' * 100);
      final truncated = Uint8List.sublistView(compressed.data, 0, compressed.data.length ~/ 2);

      expect(
        () => converter.decompressBytes(truncated),
        throwsA(
          isA<FormatException>()
              .having((e) => e.message, 'message', contains('GZip decompression failed')),
        ),
      );
    });

    test('decompressing garbage base64 throws a clean FormatException', () {
      expect(
        () => converter.decompressBase64('not valid base64 @@@@'),
        throwsA(isA<FormatException>()),
      );
    });

    test('empty input is rejected on both sides', () {
      expect(() => converter.compressText(''), throwsA(isA<ArgumentError>()));
      expect(() => converter.compressBytes(Uint8List(0)), throwsA(isA<ArgumentError>()));
      expect(() => converter.decompressBytes(Uint8List(0)), throwsA(isA<ArgumentError>()));
      expect(() => converter.decompressBase64('   '), throwsA(isA<ArgumentError>()));
    });

    test('an out-of-range compression level is rejected', () {
      expect(
        () => converter.compressBytes(Uint8List.fromList([1, 2, 3]), level: 12),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('decompressed binary that is not UTF-8 yields null text instead of throwing', () {
      final bytes = Uint8List.fromList([0xFF, 0xFE, 0x00, 0x80, 0xC0]);
      final restored = converter.decompressBytes(converter.compressBytes(bytes).data);

      expect(restored.data, bytes);
      expect(restored.textOrNull, isNull);
    });
  });

  group('execute() port', () {
    test('routes through the IToolUseCase entry point', () {
      final source = Uint8List.fromList(utf8.encode('port test'));

      final compressed = converter.execute(
        GzipConversionInput(data: source, operation: GzipOperation.compress),
      );
      final restored = converter.execute(
        GzipConversionInput(data: compressed.data, operation: GzipOperation.decompress),
      );

      expect(restored.data, source);
    });
  });
}

String _wrap(String text, int width) {
  final buffer = StringBuffer();
  for (var i = 0; i < text.length; i += width) {
    if (i > 0) buffer.write('\n');
    buffer.write(text.substring(i, i + width > text.length ? text.length : i + width));
  }
  return buffer.toString();
}

File: app/src/core/utility/gzipConverter.test.ts
import { describe, expect, it } from 'vitest'
import { GzipConverter, looksLikeGzip } from './gzipConverter'

function wrap(text: string, width: number): string {
  const lines: string[] = []
  for (let i = 0; i < text.length; i += width) {
    lines.push(text.slice(i, i + width))
  }
  return lines.join('\n')
}

describe('GzipConverter', () => {
  const converter = new GzipConverter()

  describe('compress / decompress round trip', () => {
    it('text survives a gzip round trip unchanged', () => {
      const original = 'InfraKit Studio — gzip round trip with ünïcode ✓\nline two\n'

      const compressed = converter.compressText(original)
      const restored = converter.decompressBytes(compressed.data)

      expect(restored.textOrNull).toBe(original)
      expect(restored.originalSize).toBe(new TextEncoder().encode(original).length)
    })

    it('arbitrary binary bytes survive a round trip unchanged', () => {
      const bytes = Uint8Array.from(Array.from({ length: 512 }, (_, i) => (i * 37) % 256))

      const compressed = converter.compressBytes(bytes)
      const restored = converter.decompressBytes(compressed.data)

      expect(restored.data).toEqual(bytes)
    })

    it('gzip output starts with the gzip magic number', () => {
      const compressed = converter.compressText('hello')
      expect(looksLikeGzip(compressed.data)).toBe(true)
      expect(compressed.data[0]).toBe(0x1f)
      expect(compressed.data[1]).toBe(0x8b)
    })
  })

  describe('size reporting', () => {
    it('a highly repetitive string compresses to a fraction of its size', () => {
      const repetitive = 'the same line over and over\n'.repeat(400)

      const result = converter.compressText(repetitive)

      expect(result.originalSize).toBe(new TextEncoder().encode(repetitive).length)
      expect(result.compressedSize).toBeLessThan(Math.floor(result.originalSize / 10))
      expect(result.compressionRatio).toBeLessThan(0.1)
      expect(result.spaceSavingPercent).toBeGreaterThan(90)
    })

    it('sizes are reported from the payload side regardless of direction', () => {
      const compressed = converter.compressText('a'.repeat(1000))
      const restored = converter.decompressBytes(compressed.data)

      expect(restored.originalSize).toBe(1000)
      expect(restored.compressedSize).toBe(compressed.compressedSize)
      expect(restored.compressionRatio).toBeCloseTo(compressed.compressionRatio, 9)
    })
  })

  describe('base64 transport', () => {
    it('text -> gzip -> base64 -> gzip -> text round trips', () => {
      const original = 'kubectl get pods -A | grep CrashLoopBackOff'

      const compressed = converter.compressText(original)
      const restored = converter.decompressBase64(compressed.base64)

      expect(restored.textOrNull).toBe(original)
    })

    it('base64 pasted with line breaks and no padding still decodes', () => {
      const text = 'wrapped payload '.repeat(20)
      const compressed = converter.compressText(text)

      const wrapped = wrap(compressed.base64, 40)
      const unpadded = wrapped.replace(/=/g, '')

      expect(converter.decompressBase64(wrapped).textOrNull).toBe(text)
      expect(converter.decompressBase64(unpadded).textOrNull).toBe(text)
    })

    it('url-safe base64 alphabet is accepted', () => {
      const compressed = converter.compressText('url safe alphabet test payload')
      const urlSafe = compressed.base64.replace(/\+/g, '-').replace(/\//g, '_')

      expect(converter.decompressBase64(urlSafe).textOrNull).toBe('url safe alphabet test payload')
    })
  })

  describe('error handling', () => {
    it('decompressing non-gzip bytes throws a clean error', () => {
      expect(() =>
        converter.decompressBytes(new TextEncoder().encode('this is just text')),
      ).toThrow(/Not a GZip stream/)
    })

    it('decompressing a truncated gzip stream throws a clean error', () => {
      const compressed = converter.compressText('truncate me '.repeat(100))
      const truncated = compressed.data.subarray(0, Math.floor(compressed.data.length / 2))

      expect(() => converter.decompressBytes(truncated)).toThrow(/GZip decompression failed/)
    })

    it('decompressing garbage base64 throws a clean error', () => {
      expect(() => converter.decompressBase64('not valid base64 @@@@')).toThrow()
    })

    it('empty input is rejected on both sides', () => {
      expect(() => converter.compressText('')).toThrow()
      expect(() => converter.compressBytes(new Uint8Array(0))).toThrow()
      expect(() => converter.decompressBytes(new Uint8Array(0))).toThrow()
      expect(() => converter.decompressBase64('   ')).toThrow()
    })

    it('an out-of-range compression level is rejected', () => {
      expect(() => converter.compressBytes(Uint8Array.from([1, 2, 3]), 12)).toThrow()
    })

    it('decompressed binary that is not UTF-8 yields null text instead of throwing', () => {
      const bytes = Uint8Array.from([0xff, 0xfe, 0x00, 0x80, 0xc0])
      const restored = converter.decompressBytes(converter.compressBytes(bytes).data)

      expect(restored.data).toEqual(bytes)
      expect(restored.textOrNull).toBeNull()
    })
  })

  describe('execute() port', () => {
    it('routes through the IToolUseCase entry point', () => {
      const source = new TextEncoder().encode('port test')

      const compressed = converter.execute({ data: source, operation: 'compress' })
      const restored = converter.execute({ data: compressed.data, operation: 'decompress' })

      expect(restored.data).toEqual(source)
    })
  })
})
